src/Analyzers/CSharp/Analyzers/MakeStructReadOnly/CSharpMakeStructReadOnlyDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,112 @@
+            if (!IsCandidate(context, out var typeDeclaration, out var option))
+                return;
+
+            // Looks good.  However, we have to make sure that the struct has no code which actually overwrites 'this'
+
+            var writesToThis = false;
+            context.RegisterSyntaxNodeAction(context =>
+            {
+                var semanticModel = context.SemanticModel;
+                var thisExpression = (ThisExpressionSyntax)context.Node;
+                var cancellationToken = context.CancellationToken;
+                writesToThis = writesToThis || thisExpression.IsWrittenTo(semanticModel, cancellationToken);
+            }, SyntaxKind.ThisExpression);
+
+            context.RegisterSymbolEndAction(context =>
+            {
+                // if we wrote to 'this', then we cannot convert this struct.
+                if (writesToThis)
+                    return;
+
+                context.ReportDiagnostic(DiagnosticHelper.Create(
+                    Descriptor,
+                    typeDeclaration.Identifier.GetLocation(),
+                    option.Notification.Severity,
+                    additionalLocations: ImmutableArray.Create(typeDeclaration.GetLocation()),
+                    properties: null));
+            });
+        }, SymbolKind.NamedType);//.RegisterSyntaxNodeAction(AnalyzeSyntax, SyntaxKind.StructDeclaration, SyntaxKind.RecordStructDeclaration);
+
+    private static bool IsCandidate(
+        SymbolStartAnalysisContext context,
+        [NotNullWhen(true)] out TypeDeclarationSyntax? typeDeclaration,
+        [NotNullWhen(true)] out CodeStyleOption2<bool>? option)
+    {
+        var typeSymbol = (INamedTypeSymbol)context.Symbol;
+        var cancellationToken = context.CancellationToken;
+
+        typeDeclaration = null;
+        option = null;
+        if (typeSymbol.TypeKind is not TypeKind.Struct)
+            return false;
+
+        if (typeSymbol.IsReadOnly)
+            return false;
+
+        if (typeSymbol.DeclaringSyntaxReferences.Length == 0)
+            return false;
+
+        typeDeclaration = typeSymbol.DeclaringSyntaxReferences[0].GetSyntax(cancellationToken) as TypeDeclarationSyntax;
+        if (typeDeclaration is null)
+            return false;
+
+        var options = new CSharpAnalyzerOptionsProvider(context.Options.AnalyzerConfigOptionsProvider.GetOptions(typeDeclaration.SyntaxTree), context.Options);
+        option = options.PreferReadOnlyStruct;
+        if (!option.Value)
+            return false;
+
+        // Now, ensure we have at least one field and that all fields are readonly.
+        var hasField = false;
+        foreach (var member in typeSymbol.GetMembers())
+        {
+            if (member is IFieldSymbol field)
+            {
+                hasField = true;
+                if (!field.IsReadOnly)
+                    return false;
+            }
+        }
+
+        if (!hasField)
+            return false;
+
+        return true;
```

Comment:
Consider `return hasField;`

Reply:
considered, but i prefer this approach.  it's explicit returns of 'false' in the failure paths, and true in the final success path.